src/widget/tree_structure.rs

```diff
@@ -0,0 +1,155 @@
+        let parent_id = Some(parent_id);
+        let mut old_id = None;
+
+        self.children
+            .entry(parent_id)
+            .and_modify(|children| {
+                // TODO Result instead of panic when out of bounds?
+                old_id = Some(children[idx]);
+                children[idx] = new_id;
+            })
+            .or_insert_with(|| vec![new_id]); // TODO this should not happen, Result/unwrap?
+
+        if let Some(old_id) = old_id {
+            self.parent.remove(&old_id);
+            self.parent
+                .entry(new_id)
+                .and_modify(|parent| {
+                    *parent = parent_id;
+                })
+                .or_insert(parent_id);
+        }
+    }
+
+    pub(crate) fn delete_children(&mut self, parent_id: Id, range: std::ops::Range<usize>) {
+        let parent_id = Some(parent_id);
+        let children = &self.children[&parent_id][range.clone()];
+        for child in children {
+            self.parent.remove(child);
+        }
+        self.children.get_mut(&parent_id).unwrap().drain(range);
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+
+    #[test]
+    fn mutates_simple_tree_structure() {
+        let mut tree_structure = TreeStructure::default();
+
+        let parent = Id::next();
+        let child1 = Id::next();
+        let child2 = Id::next();
+        let child3 = Id::next();
+
+        // append children
+        tree_structure.append_child(parent, child1);
+        tree_structure.append_child(parent, child2);
+        tree_structure.append_child(parent, child3);
+        let children = tree_structure.children(Some(parent)).unwrap();
+        assert_eq!(children.len(), 3);
+        assert_eq!(children[0], child1);
+        assert_eq!(children[1], child2);
+        assert_eq!(children[2], child3);
+        assert_eq!(tree_structure.parent(child1), Some(Some(parent)));
+        assert_eq!(tree_structure.parent(child2), Some(Some(parent)));
+        assert_eq!(tree_structure.parent(child3), Some(Some(parent)));
+
+        // change children
+        let child2_new = Id::next();
+        tree_structure.change_child(parent, 1, child2_new);
+        let children = tree_structure.children(Some(parent)).unwrap();
+        assert_eq!(children.len(), 3);
+        assert_eq!(children[0], child1);
+        assert_eq!(children[1], child2_new);
+        assert_eq!(children[2], child3);
+        assert_eq!(tree_structure.parent(child1), Some(Some(parent)));
+        assert_eq!(tree_structure.parent(child2), None);
+        assert_eq!(tree_structure.parent(child2_new), Some(Some(parent)));
+        assert_eq!(tree_structure.parent(child3), Some(Some(parent)));
+
+        // delete children
+        tree_structure.delete_children(parent, 0..2);
+        let children = tree_structure.children(Some(parent)).unwrap();
+        assert_eq!(children.len(), 1);
+        assert_eq!(children[0], child3);
+        assert_eq!(tree_structure.parent(child1), None);
```

Comment:
I feel like maybe all of these should panic. Maybe add an `is_in_tree()` method instead?